classes/class2/README.md

```diff
@@ -1 +1,630 @@
+### Array Methods
+
+#### Array length
+
+Retrieve the length of an array with `Array.length`:
+
+```node
+> array
+[ 1, 2, 3, 4 ]
+> array.length
+4
+```
+
+#### Appending and Prepending
+
+Use `push` to append an element to an array:
+
+```node
+> array.push(5)
+5
+> array
+[ 1, 2, 3, 4, 5 ]
+```
+
+And `pop` to remove the last element (`pop` returns the last element):
+
+```node
+> array.pop()
+5
+> array
+[ 1, 2, 3, 4 ]
+```
+
+Conversely, use `unshift` to prepend an element:
+
+```node
+> array.unshift(0)
+5
+> array
+[ 0, 1, 2, 3, 4 ]
+```
+
+And `shift` to remove the first element (`shift` returns the first element):
+
+```node
+> array.shift()
+0
+> array
+[ 1, 2, 3, 4 ]
+```
+
+#### Further Reading
+
+There are more built-in array methods, but we'll cover them later. JavaScript arrays generally act like Python arrays.
+
+Two good sources for documentation are [w3schools](http://www.w3schools.com/jsref/jsref_obj_array.asp) (easier to follow) and [Mozilla](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array) (official and more complete).
+
+### Weird JavaScript Array Behavior (Optional)
+
+Since JavaScript arrays are objects, they kind of act like them:
+
+```node
+> jsArray
+[ true, 'string', [Function] ]
+> jsArray.question = 'Am I an object?'
+'Am I an object?'
+> jsArray
+[ true,
+'string',
+[Function],
+question: 'Am I an object?' ]
+```
+
+That string was actually stored as a property, not an element of the array:
+
+```node
+> jsArray[3]
+undefined
+> jsArray.question
+'Am I an object?'
+```
+
+Arrays are, in fact, objects, but they're special. When you index an array with a non-negative integer, it acts like an array. Indexing an object similarly just casts the integer to a string and adds it as a key.
+
+```node
+> typeof(jsArray)
+'object'
+> jsArray instanceof Array // Array is a type of object
+true
+> typeof(obj)
+'object'
+> obj[4] = 6
+6
+> obj
+{ '4': 6,
+aBool: true,
+aString: 'string',
+aNum: 5.6,
+aFunc: [Function],
+anObj: { key: 'nesting' } }
+```
+
+So if you index an array with an invalid number, it won't throw an error.
+
+```node
+> jsArray[-1] = -2
+-2
+> jsArray[3.141] = 'pi'
+'pi'
+> jsArray[6] = 5
+5
+> jsArray
+[ true,
+'string',
+[Function],
+,
+,
+,
+5,
+question: 'Am I an object?',
+'-1': -2,
+'3.141': 'pi' ]
+```
+
+The moral of the story is that JavaScript is remarkably robust, whether you like it or not. It will run happily where other languages would choke, so it often falls to you to write good code that will prevent weird things from happening. Leverage the flexibility of JavaScript to write elegant code — don't let it bite you.
+
+## Homework for next class:
+
+Read the [node beginner book](http://www.nodebeginner.org) (much shorter than an actual book). This resource ties together a lot of what we covered in class 1 by implementing a basic web server in Node.js. You'll gain an understanding of how Express works while flexing your newly acquired JavaScript muscle.
```

Comment:
CHANGE: Read the free preview of the node beginner book. (So they don't get confused when it ends abruptly.)
ADD: All of the code can be run directly in node from the command line, so be sure to try out all of the exampls and understand what's happening.
CHANGE: You'll gain an understanding of what Express does internally while flexing ...